jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/lock/SelfReleaseRuntimeManagerLock.java

```diff
@@ -0,0 +1,177 @@
+            if (c == 0) {
+                if (compareAndSetState(0, acquires)) {
+                    setExclusiveOwnerThread(current);
+                    return true;
+                }
+            }
+            else if (current == getExclusiveOwnerThread()) {
+                int nextc = c + acquires;
+                if (nextc < 0) // overflow
+                    throw new Error("Maximum lock count exceeded");
+                setState(nextc);
+                return true;
+            }
+            return false;
+        }
+
+        protected final boolean tryRelease(int releases) {
+            int c = getState() - releases;
+            boolean free = false;
+            if (c == 0) {
+                free = true;
+                setExclusiveOwnerThread(null);
+            }
+            setState(c);
+            return free;
+        }
+
+        final int getHoldCount() {
+            return isHeldExclusively() ? getState() : 0;
+        }
+
+        final boolean isLocked() {
+            return getState() != 0;
+        }
+
+        final void lock() {
+            if (compareAndSetState(0, 1))
+                setExclusiveOwnerThread(Thread.currentThread());
+            else
+                acquire(1);
+        }
+
+        protected final boolean tryAcquire(int acquires) {
+            return nonfairTryAcquire(acquires);
+        }
+
+        protected final boolean isHeldExclusively() {
+            return getExclusiveOwnerThread() == Thread.currentThread();
+        }
+    }
+
+    private SelfRelaseCapableSync sync;
+    private List<StackTraceElement[]> traces;
+    private long currentThreadId;
+
+    public SelfReleaseRuntimeManagerLock() {
+        sync = new SelfRelaseCapableSync();
+        traces = new ArrayList<>();
+        currentThreadId = -1;
+    }
+    
+    @Override
+    public void lock() {
+        sync.lock();
+        addTraces();
+    }
+
+    @Override
+    public boolean tryLock(long units, TimeUnit timeUnit) throws InterruptedException {
+        boolean outcome = sync.tryAcquireNanos(1, timeUnit.toMillis(units));
+        if(outcome) {
+            addTraces();
+        }
+        return outcome;
+    }
+
+    @Override
+    public void lockInterruptible() throws InterruptedException {
+        sync.acquireInterruptibly(1);
+        addTraces();
+    }
+
+    private void addTraces() {
+        long newThreadId = Thread.currentThread().getId();
+        if(currentThreadId < 0 || currentThreadId != newThreadId) {
+            currentThreadId = newThreadId;
+            traces.clear();
+        }
+        // we add all stack traces during reentrant lock
+        traces.add(Thread.currentThread().getStackTrace());
+    }
+    @Override
+    public void unlock() {
+        // only we clear up when count is 0
+        if(currentThreadId == Thread.currentThread().getId() && sync.getHoldCount() == 1) {
+            traces.clear(); // clean up
+        }
+        sync.release(1);
```

Comment:
Shouldn't currentThreadId be set to -1 also here?

Reply:
you release 1 not ... -1

Reply:
I meant `currentThreadId = -1;`